### PR TITLE
Preserve literals across .asTypeOf

### PR DESCRIPTION
**Pull request by jackkoenig**
Built on top of https://github.com/chipsalliance/chisel/pull/4160 so should merge this one after that one (might need rebase but git seems smarter about cherry-picks these days).

This is very valuable for ChiselSim-style testing because it means that you can much more easily construct literals in tests.

### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- API modification

#### Desired Merge Strategy

- Squash

#### Release Notes

Casting a literal (of any type) to another type with .asTypeOf will result in a literal of the new type. For non-literals, the FIRRTL representation will now be a little bit more efficient.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.6.x`, `5.x`, or `6.x` depending on impact, API modification or big change: `7.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

**Comment by jackkoenig** [comment]
I was hoping to backport this because I think it's potentially a very useful feature for ChiselSim and 6.x is closer to 3.6.x which makes for easier porting, but alas it seems that it is a really common pattern to use `0.U.asTypeOf(new MyBundle)` to initialize a Wire of type `MyBundle` that then as a few fields connected to.

This change is still good and worth making, but it needs to be on a major version since it will break those uses (they just need to add a `WireInit`).

**Comment by jackkoenig** [comment]
Converted to draft temporarily. I want to deprecate assigning to the wire returned by `.asTypeOf` (and backport to 6.x) before merging this. I am almost done with the PR doing that deprecation (which involves adding my long desired `.readOnly` feature!), should open it tomorrow.

**Comment by jackkoenig** [comment]
This is ready to go

File: src/test/scala/chiselTests/stage/WarningConfigurationSpec.scala [hunk 1]
```diff
@@ -380,10 +380,10 @@ class WarningConfigurationSpec extends AnyFunSpec with Matchers with chiselTests
...
     }
 
-    it("should number AsTypeOfReadOnly as 8") {
-      val args = Array("--warn-conf", "id=8:e,any:s", "--throw-on-first-error")
+    it("should now error on AsTypeOfReadOnly") {
+      val args = Array("--throw-on-first-error")
       val e = the[Exception] thrownBy ChiselStage.emitCHIRRTL(new AsTypeOfReadOnly, args)
-      e.getMessage should include("[W008] Return values of asTypeOf will soon be read-only")
+      e.getMessage should include("Return values of asTypeOf are now read-only")
     }
```
Comment (jackkoenig) [inline src/test/scala/chiselTests/stage/WarningConfigurationSpec.scala:387]: So this is the first time we'll be elevating a warning to an error. I haven't really worked through it all, we could add a corresponding `E008` for the old `W008`, but no other errors are numbered yet (and of course, unlike warnings, you will not be allowed to configure errors). Any thoughts?

Comment (mwachs5) [reply src/test/scala/chiselTests/stage/WarningConfigurationSpec.scala:387]: it seems like if we want to add error numbers we should do that more holistically? separately from this PR?

File: core/src/main/scala/chisel3/ChiselEnum.scala [hunk 2]
```diff
@@ -73,6 +68,27 @@ abstract class EnumType(private[chisel3] val factory: ChiselEnum, selfAnnotating
   def do_>=(that: EnumType)(implicit sourceInfo: SourceInfo): Bool =
     compop(sourceInfo, GreaterEqOp, that)
 
+  // This is a hack to preserve the _workaround_ for https://github.com/chipsalliance/chisel/issues/4159
```
Comment (mwachs5) [inline core/src/main/scala/chisel3/ChiselEnum.scala:71]:
```suggestion
  // This preserves the _workaround_ for https://github.com/chipsalliance/chisel/issues/4159
```

File: core/src/main/scala/chisel3/internal/package.scala [hunk 3]
```diff
@@ -67,6 +69,40 @@ package object internal {
     if (headOk) res else s"_$res"
   }
 
+  // Hack to work around https://github.com/chipsalliance/chisel/issues/4162
```
Comment (mwachs5) [inline core/src/main/scala/chisel3/internal/package.scala:72]:
```suggestion
  // Work around https://github.com/chipsalliance/chisel/issues/4162
```

**Review by mwachs5 — APPROVED** [review]

**Pull request merged**